### PR TITLE
Add Skai migration post

### PR DESCRIPTION
**Pull request by cnunciato**
## Description

Add guest post from Danny Zalkind.

## Checklist:

- [x] I have reviewed the [style guide](https://github.com/pulumi/pulumi-hugo/blob/master/STYLE-GUIDE.md).
- [x] If blogging, I have reviewed the [blogging guide](https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md).
- [x] I have manually confirmed that all new links work.
- [x] I added aliases (i.e., redirects) for all filename changes.
- [x] If making css changes, I rebuilt the bundle.

### CODE REVIEW

**Review by dannyz1983 — APPROVED** [review]

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+authors:
+    - danny-zalkind
+tags:
+    - aws
+    - cassandra
+    - data-warehouse
+    - migration
+---
+
+> Danny Zalkind is the DevOps group manager for Skai, an award-winning intelligent marketing platform. He brings his 15 years of exprience of managing tech teams to his current role where he's dedicated to allow Skai R&D to efficiently produce and serve software. You can find him on [Linkedin](https://www.linkedin.com/in/danny-zalkind-01602b56/).
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:15]:
```suggestion
> Danny Zalkind is the DevOps group manager for Skai, an award-winning intelligent marketing platform. He brings his 15 years of experience of managing tech teams to his current role where he's dedicated to allow Skai R&D to efficiently produce and serve software. You can find him on [Linkedin](https://www.linkedin.com/in/danny-zalkind-01602b56/).
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+As Skai continues its journey towards [fully migrating to the cloud using Pulumi](https://www.pulumi.com/blog/kenshoo-migrates-to-aws-with-pulumi/), we've taken another large bite out of the migration pie, moving our most critical data to AWS on top of [Amazon Keyspaces](https://aws.amazon.com/keyspaces/), an Apache Cassandra–compatible database service.
+
+This blog post will dive into how we used Pulumi and AWS Lambda to perform the migration.
+
+## Offloading a piece of pie to AWS
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:21]:
```suggestion
## Offloading a piece of the pie to AWS
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+First, some background on what we do. Skai helps marketers work smarter and faster with a unified platform to reach and convert shoppers. An important service we provide to our clients is optimization of their digital marketing campaign efficiency. We store billions of events per day to enable clients to continuously improve the value they get from their ad spend.
+
+Our on-premises setup included a 100-node Apache Cassandra cluster to be able to withstand the scale of billions of events, some stored for prolonged periods of time. Maintaining and scaling this setup was a high effort from our Data Infra teams. It was clear to us it was time to offload that piece of the pie to our cloud provider.
+
+## Why Amazon Keyspaces
+
+When looking for a solution, we were looking for a service that will mark the following checkboxes:
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:29]:
```suggestion
When looking for a solution, we were looking for a service that would mark the following checkboxes:
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+* Fully serverless to allow us focus on our business instead of server maintenance.
+* Scalability should be automatic and easy to perform.
+* High availability and multi-region built-in.
+* Cost effective pay-as-you-go and commitment-based model.
+* Extensive support and rich documentation.
+* Fully automatable using our main Infrastructure as Code framework Pulumi.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:36]:
```suggestion
* Fully automatable using Pulumi, our main Infrastructure as Code framework.
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+## Migration method
+
+Let's take a look at how we set up the migration. We decided to migrate the data using a dual-write method, writing new data to both our on-prem Cassandra cluster and Keyspaces. Any historical data needed could be fetched from Snowflake, our data lake solution, using our Apache Spark pipeline.
+
+![Skai data-migration diagram](./skai-migration-diagram.png)
+
+Finally, to make sure that the data is valid and we have no data discrepancies on the Keyspaces side, we've added a Lambda function to periodically sample the data and compare it to the data in our on-prem cluster.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:44]:
```suggestion
Finally, to make sure that the data is valid and we have no data discrepancies on the Keyspaces side, we added a Lambda function to periodically sample the data and compare it to the data in our on-prem cluster.
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+We made a conscious decision to decouple the application code from the provisioning of the infrastructure, letting Pulumi handle all resource creation and configuration, while the application team used our standard CI/CD pipeline to deliver their code to the service owning the Keyspaces data.
+
+## Lets start provisioning
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:48]:
```suggestion
## Let's start provisioning
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+Keyspaces requires creating a logical entity called a keyspace, which groups related tables relevant for one or more applications. After that, we will iterate over our config and create each table with the relevant columns and the preferred provisioning mode --- either "on-demand" or "provisioned" depending on our workload pattern. A provisioned workload in pre-allocated and thus significantly (x7) lower in hourly price, suitable mainly for use cases where the workload is quite predictable with no spikes.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:50]:
```suggestion
Keyspaces requires creating a logical entity called a keyspace, which groups related tables relevant for one or more applications. After that, we iterate over our config and create each table with the relevant columns and the preferred provisioning mode --- either "on-demand" or "provisioned," depending on our workload pattern. A provisioned workload in pre-allocated and thus significantly (x7) lower in hourly price, suitable mainly for use cases where the workload is quite predictable with no spikes.
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+Finally we will populate AWS tags in order to be able track costs and performance metrics.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:52]:
```suggestion
Finally we populate AWS tags in order to be able track costs and performance metrics.

The following Python code illustrates the process:
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+            ),
+            opts=pulumi.ResourceOptions(
+                parent=keyspace, provider=provider, protect=True
+            ),
+        )
+```
+
+## Comparison Lambda
+
+Next we will define and create a Lambda function that samples and compares data between the data newly ingested into  Keyspaces, our on-premise Cassandra cluster, and our data lake Snowflake.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:161]:
```suggestion
Next we define and create a Lambda function that samples and compares data between the data newly ingested into  Keyspaces, our on-premise Cassandra cluster, and our data lake Snowflake.
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+We created an abstract class implementation of AWS Lambda to standardize our various Lambda implementations and keep the code readable. This class returns all needed arguments, configuration, layers, etc.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:163]:
```suggestion
We created an abstract class implementation of AWS Lambda to standardize our various Lambda implementations and keep the code readable. This class returns all needed arguments, configuration, layers, etc. Here is some Python code that shows our abstraction, named `AwsFunction`:
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+                variables=self.args.environment_variables.to_dict()
+            ),
+            vpc_config=self.args.vpc_config,
+            opts=ResourceOptions(
+                provider=self.args.provider, depends_on=[self.kms_key] + layers
+            ),
+        )
+```
+
+Now we will instantiate the abstract class as part of our "building block" component class, create the function, triggers and invocation:
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:212]:
```suggestion
Now we instantiate the abstract class as part of our "building block" component class, create the function, triggers and invocation:
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+```python
+from pulumi_komponents.awslambda import AwsLambda, AwsLambdaArgs
+from pulumi import Config
+
+config = Config()
+
+AwsLambda(AwsLambdaArgs(config))
+```
+
+Note: This only a partial code example, meant to demonstrate the way we use object-oriented paradigms to create reusable building-blocks of code to be consumed by various platform and R&D teams to provision AWS resources.
```
Comment (scottslowe) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:267]: Should we consider using `{{% notes %}}` to set this off?

**Review by scottslowe — CHANGES_REQUESTED** [review]
With the exception of the spelling error in the opening paragraph, consider the rest of these as suggestions.

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 13]
```diff
@@ -0,0 +1,279 @@
+---
+title: How Skai Migrated to Amazon Keyspaces with Pulumi
+date: 2023-05-15
```
Comment (cnunciato) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:3]:
```suggestion
date: 2023-05-16
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+authors:
+    - danny-zalkind
+tags:
+    - aws
+    - cassandra
+    - data-warehouse
+    - migration
+---
+
+> Danny Zalkind is the DevOps group manager for Skai, an award-winning intelligent marketing platform. He brings his 15 years of experience of managing tech teams to his current role where he's dedicated to allow Skai R&D to efficiently produce and serve software. You can find him on [Linkedin](https://www.linkedin.com/in/danny-zalkind-01602b56/).
```
Comment (cnunciato) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:15]:
```suggestion
> Danny Zalkind is the Senior Director of Infrastructure Engineering for Skai, an award-winning intelligent marketing platform. He brings his 15 years of experience of managing tech teams to his current role where he's dedicated to allow Skai R&D to efficiently produce and serve software. You can find him on [Linkedin](https://www.linkedin.com/in/danny-zalkind-01602b56/).
```

File: themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+date: 2023-05-15
+meta_desc: "Learn how Skai used Pulumi to manage the migration of its its on-prem Cassandra cluster to Amazon Keyspaces."
+meta_image: meta.png
+authors:
+    - danny-zalkind
+tags:
+    - aws
+    - cassandra
+    - data-warehouse
+    - migration
```
Comment (cnunciato) [inline themes/default/content/blog/how-skai-migrated-to-aws-keyspaces-with-pulumi/index.md:12]:
```suggestion
    - migration
    - guest-post
```

**Review by cnunciato — COMMENTED** [review]
Fix date, add tag.

**Pull request merged**